lib/ace/ext/easymotion.js

```diff
@@ -0,0 +1,243 @@
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+define(function(require, exports, module) {
+"use strict";
+
+var Range = require("ace/range").Range;
+
+var findings = {};
+var nextId = 0;
+
+// Vimium linkHints functionality borrowed from
+// https://github.com/philc/vimium/blob/0bf605a934/content_scripts/link_hints.coffee#L361-L398
+function logXOfBase(x, base) {
+  return Math.log(x) / Math.log(base);
+}
+
+var linkHintCharacters = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n",
+                          "o", "u"];
+
+function hintStrings(linkCount) {
+  var digitsNeeded = Math.ceil(logXOfBase(linkCount, linkHintCharacters.length));
+
+  var shortHintCount = Math.floor(
+    (Math.pow(linkHintCharacters.length, digitsNeeded) - linkCount) /
+    linkHintCharacters.length);
+
+  var longHintCount = linkCount - shortHintCount;
+
+  var hintStrings = [];
+
+  if (digitsNeeded > 1) {
+    var i;
+    for (i = 0; i < shortHintCount; i++) {
+      hintStrings.push(numberToHintString(i, linkHintCharacters, digitsNeeded - 1));
+    }
+  }
+
+  var start = shortHintCount * linkHintCharacters.length;
+  var i;
+  for (i = 0; i < start + longHintCount; i++) {
+    hintStrings.push(numberToHintString(i, linkHintCharacters, digitsNeeded))
+  }
+
+  return hintStrings;
+}
+
+function numberToHintString(number, characterSet, numHintDigits) {
+  numHintDigits = numHintDigits || 0;
+  var base = characterSet.length;
+  var hintString = [];
+  var remainder = 0;
+
+  while (true) {
+    remainder = number % base;
+    hintString.unshift(characterSet[remainder]);
+    number -= remainder;
+    number /= Math.floor(base);
+    if (number <= 0)
+      break;
+  }
+
+  var hintStringLength = hintString.length;
+  var i;
+  for (i = 0; i < numHintDigits - hintStringLength; i++) {
+    hintString.unshift(characterSet[0])
+  }
+
+  return hintString.join("");
+}
+
+function easymotionMark(editor, char) {
+    var cursor = editor.getCursorPosition(),
+        lastVisibleRow = editor.session.getScreenLength();
+
+    var text = editor.session.getTextRange(new Range(cursor.row, cursor.column, lastVisibleRow, 0));
+
+    var currentRow = cursor.row,
+        currentColumn = cursor.column;
+
+    text = text.slice(1, text.length);
+
+    nextId = 0;
+    findings = {};
+
+    var occurrences = (text.match(new RegExp(char, "g")) || []).length;
+    var hints = hintStrings(occurrences);
+
+    while (text.length > 0) {
+        currentColumn = currentColumn + 1;
```

Comment:
Using getMatchOffsets similar to https://github.com/ajaxorg/ace/blob/master/lib/ace/search_highlight.js#L63
might make this code simpler. Currently it gets confused by empty lines.